src/xen/xen.c

```diff
@@ -0,0 +1,847 @@
+    xen_hvm_param.domid = DOMID_SELF;
+    xen_hvm_param.index = HVM_PARAM_STORE_EVTCHN;
+    rv = HYPERVISOR_hvm_op(HVMOP_get_param, &xen_hvm_param);
+    if (rv < 0) {
+        msg_err("failed to get xenstore event channel (rv %d)\n", rv);
+        return false;
+    }
+
+    xen_info.xenstore_evtchn = xen_hvm_param.value;
+    xen_debug("event channel %ld, allocating and mapping shared info page", xen_info.xenstore_evtchn);
+
+    /* shared info page - taking page from identity heap, but could be backed as well */
+    xen_info.shared_info = allocate_zero(heap_pages(kh), PAGESIZE);
+    assert(xen_info.shared_info != INVALID_ADDRESS);
+    xen_add_to_physmap_t xatp;
+    xatp.domid = DOMID_SELF;
+    xatp.space = XENMAPSPACE_shared_info;
+    xatp.idx = 0;
+    xatp.gpfn = u64_from_pointer(xen_info.shared_info) >> PAGELOG; /* identity heap, v == p */
+    rv = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
+    if (rv < 0) {
+        msg_err("failed to add shared info map (rv %d)\n", rv);
+        goto out_dealloc_shared_page;
+    }
+    xen_debug("shared info page: 0x%lx", u64_from_pointer(xen_info.shared_info));
+
+    if (!xen_feature_supported(XENFEAT_hvm_callback_vector)) {
+        msg_err("HVM callback vector must be supported; xen setup failed (features mask 0x%x)\n",
+                xen_info.features);
+        goto out_dealloc_shared_page;
+    }
+
+    /* set up interrupt handling path */
+    int irq = allocate_interrupt();
+    xen_debug("interrupt vector %d; registering", irq);
+    register_interrupt(irq, closure(xen_info.h, xen_interrupt));
+
+    xen_hvm_param.domid = DOMID_SELF;
+    xen_hvm_param.index = HVM_PARAM_CALLBACK_IRQ;
+    xen_hvm_param.value = (2ull << 56) | irq;
+    rv = HYPERVISOR_hvm_op(HVMOP_set_param, &xen_hvm_param);
+    if (rv < 0) {
+        msg_err("failed to register event channel interrupt vector (rv %d)\n", rv);
+        goto out_unregister_irq;
+    }
+
+    /* register VCPU info */
+    xen_debug("registering VCPU info");
+    struct vcpu_register_vcpu_info vrvi;
+    u64 vci_pa = u64_from_pointer(xen_info.shared_info->vcpu_info); /* identity, pa == va */
+    vrvi.mfn = vci_pa >> PAGELOG;
+    vrvi.offset = vci_pa & (PAGESIZE - 1);
+    rv = HYPERVISOR_vcpu_op(VCPUOP_register_vcpu_info, 0 /* vcpu0 */, &vrvi);
+    if (rv < 0) {
+        msg_err("failed to register vcpu info (rv %d)\n", rv);
+        goto out_unregister_irq;
+    }
+
+    xen_info.evtchn_handlers = allocate_vector(xen_info.h, 1);
+    assert(xen_info.evtchn_handlers != INVALID_ADDRESS);
+
+    /* timer setup */
+    xen_debug("stopping periodic tick timer...");
+    /* attempt to disable periodic (tick) timer; won't work in older Xens... */
+    rv = HYPERVISOR_vcpu_op(VCPUOP_stop_periodic_timer, 0 /* vcpu0 */, 0);
+    if (rv < 0) {
+        msg_err("unable to stop periodic timer (rv %d)\n", rv);
+        goto out_unregister_irq;
+    }
+// XXX - disabled due to incorrect behavior on t2
```

Comment:
in what way is the behavior incorrect?

Reply:
The PV timer appears to work fine on a local Xen instance, however on t2 I'm seeing that the event channel interrupt fires well before the timer expiry set (well beyond TIMER_SLOP as defined in Linux). Since we can use hpet for the time being, I decided to address this in a follow-up PR once I've debugged it.